### ISSUE TITLE
Identifier 'a' has already been declared is not properly reported

### ISSUE THREAD

**Issue by mattcollier** [eslint/eslint#9996]
**Tell us about your environment**
Debian 9

Here is a gist of the whole affair: https://gist.github.com/mattcollier/1271484a58d5fdc623a3374242c4de30

* **ESLint Version:**
4.18.0

* **Node Version:**
8.9.4

* **npm Version:**
5.6.-

**What parser (default, Babel-ESLint, etc.) are you using?**
None

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "es6": true,
    "node": true
  },
  "parserOptions": {
    "ecmaVersion": 6
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
function y(callback) {
  setTimeout(() => callback(), 0);
}

function x(a, callback) {
  console.log('STEP 1', a);
  y(() => {
    console.log('STEP 2', a);
    const {a} = () => ({a: 5});
    callback(null, a);
  });
}

x(4, (err, result) => {
  console.log('ERROR', err);
  console.log('RESULT', result);
});

```

```bash
eslint *
```

**What did you expect to happen?**
In the following code, eslint properly reports: `Parsing error: Identifier 'a' has already been declared`, but in the aforementioned code, no error is detected
```js
function x(a, callback) {
  console.log('STEP 1', a);
  const {a} = () => ({a: 5});
  callback(null, a);
}

x(4, (err, result) => {
  console.log('ERROR', err);
  console.log('RESULT', result);
});

```

**What actually happened? Please include the actual, raw output from ESLint.**
no error detected when there should be an error: `Parsing error: Identifier 'a' has already been declared`

**Comment by gyandeeps** [comment]
Its because in `y` callback function scope, `a` is only declared once.

```js
function x(a, callback) {
  console.log('STEP 1', a);
  y(() => {
    console.log('STEP 2', a);
    const {a} = () => ({a: 5}); // first declaration of a
    callback(null, a);
  });
}
```

**Comment by mattcollier** [comment]
`a` is in scope and node reports an error
```
STEP 1 4
/home/matt/dev/sample/index.js:8
    console.log('STEP 2', a);
                          ^

ReferenceError: a is not defined
    at y (/home/matt/dev/sample/index.js:8:27)
    at Timeout.setTimeout [as _onTimeout] (/home/matt/dev/sample/index.js:2:20)
    at ontimeout (timers.js:475:11)
    at tryOnTimeout (timers.js:310:5)
    at Timer.listOnTimeout (timers.js:270:5)
```

**Comment by mattcollier** [comment]
Also, I don't know what linter is under the hood here, but an error is reported as expected: https://repl.it/repls/MiserlyPossibleDatabases

**Comment by not-an-aardvark** [comment]
In the first example, the variable `a` in the callback is shadowing the `a` parameter in the callback scope. This is valid syntax, although potentially a bit confusing to read. In the second example,  the variable `a` is redeclared in the same scope as the parameter, which is a syntax error.

If you want an error to be reported for `a` being used before it was declared, you should use the `no-use-before-define` rule.

**Comment by mattcollier** [comment]
@not-an-aardvark my expectation is that eslint is going to help prevent me from writing code that doesn't execute.

In this case, it's not catching an error on code that does not execute.

**Comment by mattcollier** [comment]
I just ran that replit on Firefox and it gives a much more accurate console error `can't access lexical declaration `a' before initialization`

So this clearly points to the use of the `no-use-before-define` rule you pointed out.  I still feel like this should be reported by default, without adding a rule.

**Comment by mattcollier** [comment]
I have added the `no-use-before-define` rule and it produces *an* error.  Not the error or the line where I think the error should occur since it does not align with where the JS engine throws, but I will not object if this issue is closed because eslint is working as intended.  Thanks for the help!

**Comment by not-an-aardvark** [comment]
Closing because this is working as intended.
